### ISSUE TITLE
Incorrect use of CRYPTOPP_BOOL_ALIGN16 in secblock.h

### ISSUE THREAD

**Issue by dwmcrobb** [weidai11/cryptopp#982]
This has been a problem for a long time (and is in 8.2.0) on some platforms.  `CRYPTOPP_BOOL_ALIGN16` is always defined, yet there are preprocessor conditionals checking if it's defined instead of using its value.

And for the final `#else`, don't assume 4 byte alignment; follow the intent of the comment.

First-pass patch, which I've been using for quite some time for builds with all ASM disabled.

```diff
--- secblock.h.orig     2020-12-08 14:05:04.000000000 -0500
+++ secblock.h  2020-12-08 23:26:20.332226000 -0500
@@ -486,7 +486,7 @@
 
 private:
 
-#if defined(CRYPTOPP_BOOL_ALIGN16) && (defined(_M_X64) || defined(__x86_64__))
+#if CRYPTOPP_BOOL_ALIGN16 && (defined(_M_X64) || defined(__x86_64__))
        // Before we can add additional platforms we need to check the
        // linker documentation for alignment behavior for stack variables.
        // CRYPTOPP_ALIGN_DATA(16) is known OK on Linux, OS X, Solaris.
@@ -497,7 +497,7 @@
        }
        CRYPTOPP_ALIGN_DATA(16) T m_array[S];
 
-#elif defined(CRYPTOPP_BOOL_ALIGN16)
+#elif CRYPTOPP_BOOL_ALIGN16
 
        // There be demons here... We cannot use CRYPTOPP_ALIGN_DATA(16)
        // because linkers on 32-bit machines (and some 64-bit machines)
@@ -539,7 +539,7 @@
 
        // CRYPTOPP_BOOL_ALIGN16 is 0. Use natural alignment of T.
        T* GetAlignedArray() {return m_array;}
-       CRYPTOPP_ALIGN_DATA(4) T m_array[S];
+       CRYPTOPP_ALIGN_DATA(sizeof(T)) T m_array[S];
 
 #endif
 
```

**Comment by noloader** [comment]
Thanks @dwmcrobb,

> This has been a problem for a long time (and is in 8.2.0) on some platforms. CRYPTOPP_BOOL_ALIGN16 is always defined, yet there are preprocessor conditionals checking if it's defined instead of using its value.

Yeah, so some background...

In the old days the `*_BOOL` macros took the value of 1 or 0. Something like:

```c++
// define CRYPTOPP_BOOL_ALIGN16 to 1 for platforms that need aligned buffers,
// like the SSE coprocessor on Intel machines.
#if __i386__ || __x86_64__
# define CRYPTOPP_BOOL_ALIGN16  1
#endif

#ifndef CRYPTOPP_BOOL_ALIGN16
# define CRYPTOPP_BOOL_ALIGN16  0
#endif
```

Later, the `*_BOOL` would be used as a template parameter. There's an example of it for `CRYPTOPP_BOOL_X86` in the old SHA1 class at [Crypto++ 5.6.2 SHA1](https://github.com/weidai11/cryptopp/blob/789f81f048c9bc472c7e7596f49b02eadd6fc1fd/sha.h#L43).

We moved away from the `BOOL` property of 1 or 0, and used it in a conventional macro way. That is, define a macro to 1 and check for `#if defined(...)`. Don't use it in template parameters. (Once we stopped using it in template parameters we no longer needed the strict bool property).

However, we left the name `*_BOOL` to avoid breaking userland code.

That's the history.

With that said, I don't think you change does any harm. It looks like you got all the uses of `CRYPTOPP_BOOL_ALIGN16 `.

**Comment by noloader** [comment]
@dwmcrobb,

Do you feel like amking a pull request. I'll get it merged today.

**Comment by noloader** [comment]
Thanks again @dwmcrobb.

This should be cleared at [Commit b08495b4ec45](https://github.com/weidai11/cryptopp/commit/b08495b4ec45).

**Comment by dwmcrobb** [comment]
I still have some systems using 5.6.2 so I'm aware of the history (since I diffed what had changed once I had crashes from the issue).

Just to be clear for the benefit of those using CRYPTOPP_DISABLE_ASM... the issues were not harmless.  I wouldn't have noticed nor needed to create the patch if that were the case.  Memory leaks or crashes tend to get my attention.  :-)  From recollection since it's been a while since I first created this patch for my systems... track down what happens in deallocate() and AlignedDeallocate()/UnalignedDeallocate() and you'll see where things can go awry on some systems if CRYPTOPP_BOOL_ALIGN16 is treated as defined/undefined in the patched part of secblock.h but as a value everywhere else.

I realize there is probably little testing on platforms with CRYPTOPP_DISABLE_ASM set.  Just trying to help others that might need/want to define CRYPTOPP_DISABLE_ASM (even on modern amd64 systems for whatever reason) by upstreaming the patch.

**Comment by noloader** [comment]
> Just to be clear for the benefit of those using CRYPTOPP_DISABLE_ASM...

Oh, yes. That's a problem of the library in the past. It would surface on MinGW on occasion, if I recall properly. It seems someone was distributing the library built with `CRYPTOPP_DISABLE_ASM` on MinGW. Regular users would come along and not define the macro. Then some allocation symbols would go missing.

We fixed that several years ago. Instead of tying the allocation symbols to `CRYPTOPP_DISABLE_ASM`, we time them to the platform like `__i386__` or `__x86_64__`. Now the allocation symbols are always present, regardless of `CRYPTOPP_DISABLE_ASM`.

**Comment by dwmcrobb** [comment]
Not a missing symbol issue here.  That would've been much nicer since I'd have seen the issue at link time instead of runtime.

If CRYPTOPP_DISABLE_ASM is defined, CRYPTOPP_BOOL_ALIGN16 gets defined as 0.  That's the first non-include thing in config_align.h (in master today).  It doesn't depend on platform macros in this case.  So what would happen before the patch... you'd get the 'there be demons here...' code if _M_64 and __x86_64__ were not defined.  The compiler would not see the code in the final `#else `because this was always true when CRYPTOPP_DISABLE_ASM was defined:

`#elif defined(CRYPTOPP_BOOL_ALIGN16)`

Glancing at this now... that code isn't just 'there be demons here...'.  It only works correctly on the heap if m_array happens to get 16-byte alignment (despite only requesting 8 byte alignment), yet there's nothing preventing it from being on the heap.  In fact in a lot of cases I'd expect it to be on the heap because it's inside an object that was allocated on the heap (possibly deeply nested in a class composition).  I don't think it's unreasonable to think someone might have a CryptoPP::XChaCha20::Encryption on the heap, for example (that might have been one that burned me, I switched to XChaCha20/Poly1305 for some things when I switched to 8.2.0).

Let's take an example from poly1305.h and assume a non-AIX platform:

`FixedSizeAlignedSecBlock<word32, 5> m_h;
`
PAD would be 2
For simplicity, let's choose a small heap address for m_array that is 8-byte aligned as requested: at address 0x1008 (4104).
It gets 7 elements (5+PAD), hence 28 bytes.  On the heap, that'd be bytes 0x1008 (4104) to 0x1023 (4131).
m_array % 16 = 0x08
p_array = 0x1008 - 0x08 = 0x1000.  Doesn't point to memory owned by m_array.
Both of the CRYPTOPP_ASSERTs here will pass, so the user has no idea that something is VERY wrong.

To me, this is very dangerous and fundamentally broken.  You might get lucky and m_array is 16-byte aligned, but if you don't get lucky... any instance of FixedSizeAllocatorWithCleanup<T,S,A,true> on the heap is a disaster waiting to happen.  SecureWipeArray(), for example, is going to stomp on heap we don't own.  This class template likely isn't used directly by a cryptopp user, but they will be burned by it just by putting a higher-level object on the heap that contains an instance of this template.

Of course the hard question is what's the right fix, how deep are the implications, what platforms do you want to continue to support, etc.  C++17 has a pile of alignment features that make this whole thing moot (and C++11 has enough to make it moot for this particular case), and it seems logical to move in that direction.  In the meantime it seems logical to me to make sure this code is never used.

**Comment by noloader** [comment]
> Not a missing symbol issue here. That would've been much nicer since I'd have seen the issue at link time instead of runtime.

Yeah, so how it surfaced under the MinGW problem... The MinGW author built with `-DRYPTOPP_DISABLE_ASM`. That caused the library (both *.a and *.so) to export `UnalignedAllocate` and `UnalignedDeallocate`. Regular users _did not_ build with the define, so they wanted to link against `AlignedAllocate` and `AlignedDeallocate`, which were missing. So the user's program would fail to load. No code was executed.

>  [m_array ] only works correctly on the heap if m_array happens to get 16-byte alignment (despite only requesting 8 byte alignment), yet there's nothing preventing it from being on the heap.

I have not seen a problem in practice. But if there is a problem, then we need to revisit it.

**Comment by dwmcrobb** [comment]
> > [m_array ] only works correctly on the heap if m_array happens to get 16-byte alignment (despite only requesting 8 byte alignment), yet there's nothing preventing it from being on the heap...
> 
> I have not seen a problem in practice. But if there is a problem, then we need to revisit it.
> 
> The problem area was a stack allocation, which only got 4 byte alignment on some platforms. That's the reason for the gyrations about `PAD`.

Understood.  If it were me... buh-bye old platforms.  That's always a painful decision for some, but on the other hand, C++11 is almost a decade old.  At some point you have to leave the old stuff behind when it's impeding the majority of users and isn't really doing much for the old platforms either.  They can continue to use older releases, of course, until they're ready to make a move, so it's not like they're completely abandoned.

> 
> I've never liked that stack allocation trickery. It exists today for historical purposes. I've thought about removing the stack allocation completely and always use a heap allocation. It is _a lot_ easier to get an aligned block on the heap.

Yes, the problem with the way the code is written is that it's only actually stack allocation if the template instantiation happens to be on the stack.  But it _assumes_ m_array is on the stack, which could easily be an incorrect assumption.  And the result when that assumption is wrong is potentially very bad.

In C++11, I'd argue it's easier to get an aligned block on the stack (or any auto variable) than the heap because we have `alignedas()`.  No such mechanism for anything from operator new (except placement new) until C++17.  But I think moving toward just using `alignedas()` is the smart direction.  Use language support to keep the code simpler and more flexible (as of C++17, it becomes easy to have alignment within the standard containers, etc.).  Fall back to no alignment if `alignedas()` isn't available.  And we can all yell at toolchain providers who don't correctly honor `alignedas()` in C++11 and newer toolchains.

I haven't checked... are there parts of the code asking for aligned arrays that don't need it (because they're not using any special instructions that require special alignment)?  Changing those (if there are any) to not ask for alignment they don't need might help reduce the impact of changes to users of the library.  I only mention it because I raised an eyebrow a bit when I saw `FixedSizeAlignedSecBlock` in files for ciphers I wouldn't expect to be using any special hardware acceleration.

> 
> Ping @mouse07410. What do you think about taking the `FixedSizeAllocator<T, S>`, which uses the stack, and moving them to heap allocations so we can control alignment better? (I think Wei's intent way back when was to avoid `new` and `delete` when the size of the object was known in advance, like `word32 x[5]`).

I think you still want stack allocations when the template instantiation is on the stack, for all the obvious reasons (hotness of the stack being a big one on many platforms).  `alignedas()` will get you that, and to me is the long-term solution.  Make the toolchain do the work, keep the assertion that it does the right thing.

For a fix that somewhat mimics the current code... just look for a solution that doesn't explicitly use stack addressing assumptions?  I haven't coded it, but... today the `GetAlignedArray()` we're talking about isn't bound to any must-be-known-at-compile-time constraints.  So instead of assuming stack addressing, you just need a conditional check of the addressing direction of `m_array` (compare `&m_array[0]` to `&m_array[1]`), and then set `p_array` appropriately and assert that it's in range?

**Comment by dwmcrobb** [comment]
Sigh, I'm not thinking clearly.  Can't check for heap versus stack by comparing the addresses of the array elements.  So that last paragraph doesn't make sense.

**Comment by dwmcrobb** [comment]
I was trying to picture the platform that'd have an array that grows downward without violating layout rules.  Doesn't seem possible.  I've never seen it, and it seems like if such a system existed for C or C++, it'd be documented on the web (in infamy?).

Stack _frames_ usually grow downward.  And individual variables may be re-ordered and hence be above or below each other.  But the layout of an array... contiguous and the addresses of the elements are always upward with increasing indexes.  I don't see how it could be otherwise without violating layout rules but I'm open to being educated!

Anyway, this means the code doesn't work on any platform known to me, doesn't matter whether it's on the stack or on the heap.  The good news is that it makes it easy to make it work correctly and not worry about where it lives.  It can also be done with any desired alignment with an additional template parameter.

I think this hasn't been caught before for a number of reasons:

- Not many people wind up with this code?  Perhaps the combination of macros to get here aren't common (few are compiling with `CRYPTOPP_DISABLE_ASM` defined).
- The alignment for the common cases might be 16 on the common platforms.  Hence problem not triggered even if you get here.
- The asserts don't catch it, most users aren't running debug builds anyway, and hence...
- Luck?  This one jives with many errors of this type, they can go unnoticed for a long time depending on what's being stepped on.  We've all been there.

The existing code could be corrected to get the desired alignment in a manner similar to what was done originally.  So it could just be two paths: one that uses `alignas()` and a fallback that uses a fixed version of the old code.  That fallback could be as flexible alignment-wise as `alignas()` if an additional template parameter is added, but that would seem to prompt a more holistic look at things.

I could take a crack at this in stages if you'd like, first just fixing the existing problematic case.
